### PR TITLE
Switch tests from Boost unit tests to catch2

### LINKED ISSUE #510: Use Catch2 for unit testing
I think it may be worthwhile switching to Catch2 for unit testing, as many of our newer projects have done. In the past, I've experienced bad linking issues with Boost's unit tests with the same code paths but different operating systems.

This would also allow us to make Boost fully optional and maybe even disable it by default.

### PR DESCRIPTION
**Pull request by matz-e**
Fixes #510. CTest now generates nice output, and testing error messages
should be more meaningful.

Left over:
- [x] MPI integration for the parallel tests

### CODE REVIEW

**Comment by matz-e** [comment]
@ohm314, @alkino, I've fixed the MPI invocation stuff, so this should be a full replacement for BOOST now. Should just figure out how to fix the invocation in the CI / configure OpenMPI there

File: tests/unit/tests_high_five_base.cpp [hunk 1]
```diff
@@ -1026,7 +1013,7 @@ void attribute_scalar_rw() {
     Group g = h5file.createGroup("metadata");
 
     bool family_exist = g.hasAttribute("family");
-    BOOST_CHECK(!family_exist);
+    CHECK(!family_exist);
```
Comment (ohm314) [inline tests/unit/tests_high_five_base.cpp:1016]:
```suggestion
    CHECK(!g.hasAttribute("family"));
```

**Review by ohm314 — COMMENTED** [review]
looks overall good, just had a few minor suggestions.

File: tests/unit/tests_high_five_base.cpp [hunk 2]
```diff
@@ -685,22 +670,24 @@ BOOST_AUTO_TEST_CASE(HighFiveReadWriteShortcut) {
 
     std::vector<int> result;
     dataset.read(result);
-    BOOST_CHECK_EQUAL_COLLECTIONS(vec.begin(), vec.end(), result.begin(), result.end());
+    std::vector<unsigned> result_transformed;
+    std::copy(std::begin(result), std::end(result), std::back_inserter(result_transformed));
+    CHECK_THAT(vec, Equals(result_transformed));
```
[the remark this answers is not in the document]
Comment (matz-e) [reply tests/unit/tests_high_five_base.cpp:675]: I think the boost test may have been more lenient with types

File: tests/unit/tests_high_five_base.cpp [hunk 3]
```diff
@@ -685,22 +670,24 @@ BOOST_AUTO_TEST_CASE(HighFiveReadWriteShortcut) {
 
     std::vector<int> result;
```
[the remark this answers is not in the document]
Comment (matz-e) [reply tests/unit/tests_high_five_base.cpp]: I switched this now, reads to `unsigned` now.

**Review by alkino — APPROVED** [review]

**Pull request merged**